packages/netsuite-adapter/src/change_validators/unique_fields.ts

```diff
@@ -0,0 +1,218 @@
+    if (elemID.typeName.startsWith(CUSTOM_RECORD_TYPE_NAME_PREFIX)) {
+      return 'customRecordField'
+    }
+  }
+
+  return undefined
+}
+
+const getEmptyRestrictedTypeRecord = <T> () : Record<RestrictedType, T[]> => ({
+  savedSearch: [],
+  financialLayout: [],
+  customRecordField: [],
+})
+
+const getTypeToDataRecord = <T> (
+  elements: T[],
+  getElemID: GetElemIDFunc<T>,
+  includeTypeForFields = true,
+  typesToInclude?: Set<RestrictedType>,
+):
+  Record<RestrictedType, T[]> => {
+  const typeToDataRecord = getEmptyRestrictedTypeRecord<T>()
+
+  const addToGroup = (elem: T): void => {
+    const elemID = getElemID(elem)
+    const restrictedType = getRestrictedType(elemID, includeTypeForFields)
+    if (isDefined(restrictedType)) {
+      if (isUndefined(typesToInclude) || typesToInclude.has(restrictedType)) {
+        typeToDataRecord[restrictedType].push(elem)
+      }
+    }
+  }
+
+  elements.forEach(addToGroup)
+
+  return typeToDataRecord
+}
+
+const getTypeToRestrictedFields = async (
+  elementsSource: ReadOnlyElementsSource,
+  typeToElementsRecord: Record<RestrictedType, ElemID[]>):
+  Promise<Record<RestrictedType, string[]>> => {
+  const typeToFields: Record<RestrictedType, string[]> = getEmptyRestrictedTypeRecord<string>()
+
+  const setMap = async (type: RestrictedType, elementsIDs: ElemID[]): Promise<void> => {
+    const getters = restrictedTypeGettersMap[type]
+    const fields = await awu(elementsIDs)
+      .flatMap(elemID => getters.getSourceRestrictedFields({ elemID, elementsSource }))
+      .toArray()
+    typeToFields[type] = fields
+  }
+
+  await awu(Object.entries(typeToElementsRecord)).forEach(typeToElements =>
+    setMap(typeToElements[0] as RestrictedType, typeToElements[1]))
+
+  return typeToFields
+}
+
+const validateDuplication = async (
+  changes: ChangeDataType[],
+  uniqueFieldToID: Dictionary<number>,
+  type: RestrictedType):
+  Promise<Array<ChangeError>> => {
+  const getters = restrictedTypeGettersMap[type]
+
+  return awu(changes)
+    .map(change => (
+      { elemID: change.elemID,
+        field: getters.getChangeRestrictedField(change) }))
+    .filter(({ field }) => (uniqueFieldToID[field] ?? 0) > 1)
+    .map(({ elemID, field }): ChangeError => ({
+      elemID,
+      severity: 'Error',
+      message: getters.getMessage(),
+      detailedMessage: getters.getDetailedMessage(field),
+    }))
+    .toArray()
+}
+
+const changeValidator: NetsuiteChangeValidator = async (changes, _deployReferencedElements, elementsSource) => {
+  const changeElements = changes
+    .filter(isAdditionOrModificationChange)
+    .map(getChangeData)
+
+  const typesToChanges = getTypeToDataRecord(changeElements, (element: ChangeDataType) => element.elemID)
+  const relevantTypes = new Set(Object.entries(typesToChanges)
+    .filter(typeToChanges => typeToChanges[1].length > 0)
+    .map(typeToChanges => typeToChanges[0] as RestrictedType))
```

Comment:
nit - lodash have some helpful functions
```suggestion
  const relevantTypes = new Set(
    Object.keys(_.pickBy(typesToChanges, changes => changes.length > 0) as RestrictedType[]
  )
```